File: src/chip.rs
pub enum Color {
    White,
    Orange,
    Green,
    Blue,
    Red,
    Yellow,
    Purple,
    Black,
}

#[derive(Debug, Clone)]
pub enum Chip {
    White1,
    White2,
    White3,
    Orange1,
    Green1,
    Green2,
    Green4,
    Blue1,
    Blue2,
    Blue4,
    Red1,
    Red2,
    Red4,
    Yellow1,
    Yellow2,
    Yellow4,
    Purple1,
    Black1,
}

impl Chip {
    pub fn value(&self) -> u8 {
        return match self {
            Chip::White1
            | Chip::Orange1
            | Chip::Green1
            | Chip::Blue1
            | Chip::Red1
            | Chip::Black1
            | Chip::Yellow1
            | Chip::Purple1 => 1,
            Chip::White2 | Chip::Green2 | Chip::Blue2 | Chip::Red2 | Chip::Yellow2 => 2,
            Chip::White3 => 3,
            Chip::Green4 | Chip::Blue4 | Chip::Red4 | Chip::Yellow4 => 4,
        };
    }

    pub fn color(&self) -> Color {
        return match self {
            Chip::White1 | Chip::White2 | Chip::White3 => Color::White,
            Chip::Orange1 => Color::Orange,
            Chip::Green1 | Chip::Green2 | Chip::Green4 => Color::Green,
            Chip::Blue1 | Chip::Blue2 | Chip::Blue4 => Color::Blue,
            Chip::Red1 | Chip::Red2 | Chip::Red4 => Color::Red,
            Chip::Yellow1 | Chip::Yellow2 | Chip::Yellow4 => Color::Yellow,
            Chip::Purple1 => Color::Purple,
            Chip::Black1 => Color::Black,
        };
    }
}

File: lib/chip.ml
open Base

type ingredient =
  | Black_deaths_head_hawkmoth
  | Blue_crow_skull
  | Green_garden_spider
  | Orange_pumpkin
  | Purple_ghosts_breath
  | Red_toadstool
  | White_snowberries
  | Yellow_mandrake

type t = ingredient * int

let show : t -> string = function
  | Black_deaths_head_hawkmoth, n -> Printf.sprintf "Black %d" n
  | Blue_crow_skull, n -> Printf.sprintf "Blue %d" n
  | Green_garden_spider, n -> Printf.sprintf "Green %d" n
  | Orange_pumpkin, n -> Printf.sprintf "Orange %d" n
  | Purple_ghosts_breath, n -> Printf.sprintf "Purple %d" n
  | Red_toadstool, n -> Printf.sprintf "Red %d" n
  | White_snowberries, n -> Printf.sprintf "White %d" n
  | Yellow_mandrake, n -> Printf.sprintf "Yellow %d" n

let is_same_kind : kind:ingredient -> t -> bool =
 fun ~kind chip -> phys_equal (fst chip) kind
